simulation/halsim_ws_client/src/main/native/cpp/HALSimWSClientConnection.cpp

```diff
@@ -0,0 +1,121 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2020 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "HALSimWSClientConnection.h"
+
+#include <wpi/raw_ostream.h>
+#include <wpi/raw_uv_ostream.h>
+
+#include "HALSimWSClient.h"
+
+namespace uv = wpi::uv;
+
+namespace wpilibws {
+
+void HALSimWSClientConnection::Initialize() {
+  // Get a shared pointer to ourselves
+  auto self = this->shared_from_this();
+
+  auto hws = HALSimWS::GetInstance();
+  std::string reqHost =
+      hws->GetTargetHost() + ":" + std::to_string(hws->GetTargetPort());
+
+  auto ws =
+      wpi::WebSocket::CreateClient(*m_stream, hws->GetTargetUri(), reqHost);
+
+  ws->SetData(self);
+
+  m_websocket = ws.get();
+
+  // Hook up events
+  m_websocket->open.connect_extended([this](auto conn, wpi::StringRef) {
+    conn.disconnect();
+
+    m_buffers = std::make_unique<BufferPool>();
+
+    m_exec =
+        UvExecFunc::Create(m_stream->GetLoop(), [](auto out, LoopFunc func) {
+          func();
+          out.set_value();
+        });
+
+    auto hws = HALSimWS::GetInstance();
+    if (!hws) {
+      wpi::errs() << "Unable to get hws instance\n";
+      return;
+    }
+
+    if (!hws->RegisterWebsocket(GetSharedFromThis())) {
+      wpi::errs() << "Unable to register websocket\n";
+      return;
+    }
+
+    m_ws_connected = true;
+    wpi::errs() << "HALSimWS: WebSocket Connected\n";
+  });
+
+  m_websocket->text.connect([this](wpi::StringRef msg, bool) {
+    auto hws = HALSimWS::GetInstance();
+    if (!m_ws_connected || !hws) {
+      return;
+    }
+
+    wpi::json j;
+    try {
+      j = wpi::json::parse(msg);
+    } catch (const wpi::json::parse_error& e) {
+      std::string err("JSON parse failed: ");
+      err += e.what();
+      wpi::errs() << err << "\n";
+      return;
```

Comment:
On this side of the connection, we just report the error, but in the server implementation we terminate the WS connection on a JSON parse error.  I'm thinking terminating the connection is probably the right call, but we should be consistent.